web/swift/swift.go

```diff
@@ -0,0 +1,239 @@
+
+// GetObject retrieves a Swift object from an instance
+func GetObject(c echo.Context) error {
+	domain := c.Param("domain")
+	object := c.Param("object")
+	unescaped, err := url.PathUnescape(object)
+	if err != nil {
+		return err
+	}
+
+	i, err := lifecycle.GetInstance(domain)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("test", unescaped)
+	buf := new(bytes.Buffer)
+	sc := config.GetSwiftConnection()
+	_, err = sc.ObjectGet(swiftContainer(i), unescaped, buf, false, nil)
+	if err != nil {
+		return err
+	}
+
+	out := struct {
+		Content string `json:"content"`
+	}{
+		buf.String(),
+	}
+
+	return c.JSON(http.StatusOK, out)
+}
+
+// PutObject puts an object into Swift
+func PutObject(c echo.Context) error {
+	type reqStruct struct {
+		Instance    string `json:"instance"`
+		ObjectName  string `json:"object_name"`
+		Content     string `json:"content"`
+		ContentType string `json:"content_type"`
+	}
+
+	var req reqStruct
+	err := json.NewDecoder(c.Request().Body).Decode(&req)
+	if err != nil {
+		return err
+	}
+
+	i, err := lifecycle.GetInstance(req.Instance)
+	if err != nil {
+		return err
+	}
+
+	sc := config.GetSwiftConnection()
+	f, err := sc.ObjectCreate(swiftContainer(i), req.ObjectName, true, "", req.ContentType, nil)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	r := bytes.NewReader([]byte(req.Content))
+	_, err = io.Copy(f, r)
+	if err != nil {
+		return err
+	}
+
+	return c.JSON(http.StatusOK, nil)
+
+}
+
+// DeleteObject removes an object from Swift
+func DeleteObject(c echo.Context) error {
+	domain := c.Param("domain")
+	objectName := c.Param("object")
+	unescaped, err := url.PathUnescape(objectName)
+	if err != nil {
+		return err
+	}
+
+	i, err := lifecycle.GetInstance(domain)
+	if err != nil {
+		return err
+	}
+
+	sc := config.GetSwiftConnection()
+	err = sc.ObjectDelete(swiftContainer(i), unescaped)
+	if err != nil {
+		return err
+	}
+	return c.JSON(http.StatusOK, nil)
+}
+
+// ListObjects list objects of an instance
+func ListObjects(c echo.Context) error {
+	domain := c.Param("domain")
+	i, err := lifecycle.GetInstance(domain)
+	if err != nil {
+		return err
+	}
+	sc := config.GetSwiftConnection()
+	container := swiftContainer(i)
+
+	outNames := []string{}
+
+	err = sc.ObjectsWalk(container, nil, func(opts *swift.ObjectsOpts) (interface{}, error) {
+		names, err := sc.ObjectNames(container, opts)
+		if err == nil {
+			outNames = append(outNames, names...)
+		}
+		return names, err
+	})
+	if err != nil {
+		return err
+	}
+
+	out := struct {
+		ObjectNameList []string `json:"objects_names"`
+	}{
+		outNames,
+	}
+	return c.JSON(http.StatusOK, out)
+}
+
+// Routes sets the routing for the status service
+func Routes(router *echo.Group) {
+	router.GET("/layouts", ListLayouts, checkSwift)
+	router.GET("/get/:domain/:object", GetObject, checkSwift)
+	router.POST("/put", PutObject, checkSwift)
+	router.DELETE("/:domain/:object", DeleteObject, checkSwift)
+	router.GET("/ls/:domain", ListObjects, checkSwift)
```

Comment:
- I would prefer to use the `Host` HTTP header to give the domain
- I don't like verbs in the paths (the only HTTP verbs are GET/POST/PUT/etc. and are not in the paths)
- I want to be able in the future to add some routes for debugging the global swift containers (apps, dynamic assets)

So, I suggest to use this routes:

- GET /swift/layouts (like it is already the case)
- GET /swift/vfs/:object to get an object
- PUT /swift/vfs/:object to put an object
- DELETE /swift/vfs/:object to delete an object
- GET /swift/vfs to list all the objects of an instance VFS